File: src/gui/asemanjavalayer.cpp
/*
    Copyright (C) 2017 Aseman Team
    http://aseman.co

    AsemanQtTools is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AsemanQtTools is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "asemanjavalayer.h"

#include <QDebug>
#include <QSet>
#include <QPair>
#include <QList>
#include <QDateTime>
#include <QFile>
#include <QCoreApplication>
#include <QTimer>
#include <QScreen>
#include <QGuiApplication>
#include <QPointer>
#include <QColor>

#include <jni.h>

#if (QT_VERSION >= QT_VERSION_CHECK(6, 0, 0))
#include <QJniEnvironment>
#include <QJniObject>
typedef QJniEnvironment QAndroidJniEnvironment;
typedef QJniObject QAndroidJniObject;
#else
#include <QAndroidJniEnvironment>
#include <QAndroidJniObject>
#endif

static QSet<AsemanJavaLayer*> java_layers_objects;
static QList< QPair<QString,QString> > java_layer_inc_share_buffer;
static QList< QString > java_layer_inc_image_buffer;
static QList< QString > java_layer_inc_deeplinks_buffer;

static bool aseman_jlayer_registerNativeMethods();
static bool aseman_jlayer_native_methods_registered = false;

class AsemanJavaLayerPrivate
{
public:
    QAndroidJniObject object;
    QAndroidJniEnvironment env;
};

AsemanJavaLayer::AsemanJavaLayer() :
    QObject()
{
    if(!aseman_jlayer_native_methods_registered)
        aseman_jlayer_registerNativeMethods();

    p = new AsemanJavaLayerPrivate;
    p->object = QAndroidJniObject("io/aseman/android/AsemanJavaLayer");

#if defined(Q_OS_ANDROID) && defined(ASEMAN_STATIC_BUILD)
    QMetaObject::invokeMethod( this, "registerObject", Qt::QueuedConnection );
#else
    registerObject();
#endif

    QMetaObject::invokeMethod( this, "load_buffer", Qt::QueuedConnection );

    setImplemented(true);
}

void AsemanJavaLayer::registerObject()
{
    java_layers_objects.insert(this);
}

AsemanJavaLayer *AsemanJavaLayer::instance()
{
    static QPointer<AsemanJavaLayer> java_layer_instance;
    if( !java_layer_instance )
    {
        java_layer_instance = new AsemanJavaLayer();
        if(QCoreApplication::instance())
            QObject::connect(QCoreApplication::instance(), &QCoreApplication::destroyed, java_layer_instance, &AsemanJavaLayer::deleteLater);
    }

    return java_layer_instance;
}

bool AsemanJavaLayer::sharePaper(const QString &title, const QString &msg)
{
    QAndroidJniObject jtitle = QAndroidJniObject::fromString(title.toUtf8());
    QAndroidJniObject jmsg = QAndroidJniObject::fromString(msg.toUtf8());
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Ljava/lang/String;Ljava/lang/String;)Z", jtitle.object<jstring>(), jmsg.object<jstring>() );
    return res;
}

bool AsemanJavaLayer::shareFile(const QString &path, const QString &type)
{
    QAndroidJniObject jpath = QAndroidJniObject::fromString(path.toUtf8());
    QAndroidJniObject jtype = QAndroidJniObject::fromString(type.toUtf8());
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Ljava/lang/String;Ljava/lang/String;)Z", jpath.object<jstring>(), jtype.object<jstring>() );
    return res;
}

bool AsemanJavaLayer::openFile(const QString &path, const QString &type)
{
    QAndroidJniObject jpath = QAndroidJniObject::fromString(path.toUtf8());
    QAndroidJniObject jtype = QAndroidJniObject::fromString(type.toUtf8());
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Ljava/lang/String;Ljava/lang/String;)Z", jpath.object<jstring>(), jtype.object<jstring>() );
    return res;
}

bool AsemanJavaLayer::startCamera(const QString &output)
{
    QAndroidJniObject joutput = QAndroidJniObject::fromString(output.toUtf8());
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Ljava/lang/String;)Z", joutput.object<jstring>() );
    return res;
}

bool AsemanJavaLayer::callNumber(const QString &number)
{
    QAndroidJniObject jnumber = QAndroidJniObject::fromString(number.toUtf8());
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Ljava/lang/String;)Z", jnumber.object<jstring>() );
    return res;
}

bool AsemanJavaLayer::getOpenPictures()
{
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "()Z");
    return res;
}

bool AsemanJavaLayer::startForeground(qint32 id, const QString &title, const QString &msg, const QString &iconPath, const QString &icon, const QString &channelId)
{
    jint jid = id;
    QAndroidJniObject jtitle = QAndroidJniObject::fromString(title);
    QAndroidJniObject jmsg = QAndroidJniObject::fromString(msg);
    QAndroidJniObject jicon = QAndroidJniObject::fromString(icon);
    QAndroidJniObject jiconPath = QAndroidJniObject::fromString(iconPath);
    QAndroidJniObject jchannelId = QAndroidJniObject::fromString(channelId);
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(ILjava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)Z",
                                                  jid, jtitle.object<jstring>(), jmsg.object<jstring>(),
                                                  jiconPath.object<jstring>(), jicon.object<jstring>(),
                                                  jchannelId.object<jstring>());
    return res;
}

bool AsemanJavaLayer::stopForeground(bool removeNotification)
{
    jboolean jremoveNotification = removeNotification;
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Z)Z", jremoveNotification);
    return res;
}

bool AsemanJavaLayer::startNotification(qint32 id, const QString &title, const QString &msg, const QString &iconPath, const QString &icon, const QString &channelId, bool sound, bool vibrate)
{
    jint jid = id;
    QAndroidJniObject jtitle = QAndroidJniObject::fromString(title);
    QAndroidJniObject jmsg = QAndroidJniObject::fromString(msg);
    QAndroidJniObject jicon = QAndroidJniObject::fromString(icon);
    QAndroidJniObject jiconPath = QAndroidJniObject::fromString(iconPath);
    QAndroidJniObject jchannelId = QAndroidJniObject::fromString(channelId);
    jboolean jsound = sound;
    jboolean jvibrate = vibrate;
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(ILjava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;ZZ)Z",
                                                  jid, jtitle.object<jstring>(), jmsg.object<jstring>(),
                                                  jiconPath.object<jstring>(), jicon.object<jstring>(),
                                                  jchannelId.object<jstring>(), jsound, jvibrate);
    return res;
}

bool AsemanJavaLayer::stopNotification(qint32 id)
{
    jint jid = id;
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(I)Z", jid);
    return res;
}

QString AsemanJavaLayer::createNotificationChannel(const QString &channelId, const QString &channelName, int importance)
{
    QAndroidJniObject jchannelId = QAndroidJniObject::fromString(channelId);
    QAndroidJniObject jchannelName = QAndroidJniObject::fromString(channelName);
    jint jimportance = importance;
    QString res = p->object.callObjectMethod(__FUNCTION__, "(Ljava/lang/String;Ljava/lang/String;I)Ljava/lang/String;",
                                             jchannelId.object<jstring>(), jchannelName.object<jstring>(), jimportance).toString();
    return res;
}

bool AsemanJavaLayer::transparentStatusBar()
{
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "()Z");
    return res;
}

bool AsemanJavaLayer::transparentNavigationBar()
{
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "()Z");
    return res;
}

bool AsemanJavaLayer::startService()
{
    return startQtService();
}

bool AsemanJavaLayer::stopService()
{
    return stopQtService();
}

bool AsemanJavaLayer::killService(const QString &serviceName)
{
    QAndroidJniObject jserviceName = QAndroidJniObject::fromString(serviceName.toUtf8());
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "(Ljava/lang/String;)Z", jserviceName.object<jstring>());
    return res;
}

bool AsemanJavaLayer::startQtService()
{
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "()Z" );
    return res;
}

bool AsemanJavaLayer::stopQtService()
{
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "()Z" );
    return res;
}

int AsemanJavaLayer::densityDpi()
{
    jint res = p->object.callMethod<jint>(__FUNCTION__, "()I" );
    return res;
}

int AsemanJavaLayer::screenSizeWidth()
{
    jint res = p->object.callMethod<jint>(__FUNCTION__, "()I" );
    return res;
}

int AsemanJavaLayer::screenSizeHeight()
{
    jint res = p->object.callMethod<jint>(__FUNCTION__, "()I" );
    return res;
}

int AsemanJavaLayer::getSizeName()
{
    jint res = p->object.callMethod<jint>(__FUNCTION__, "()I" );
    return res;
}

int AsemanJavaLayer::statusBarHeight()
{
    jint res = p->object.callMethod<jint>(__FUNCTION__, "()I" );
    return res;
}

int AsemanJavaLayer::navigationBarHeight()
{
    jint res = p->object.callMethod<jint>(__FUNCTION__, "()I" );
    return res;
}

bool AsemanJavaLayer::isTablet()
{
    jboolean res = p->object.callMethod<jboolean>(__FUNCTION__, "()Z");
    return res;
}

qreal AsemanJavaLayer::density()
{
    jfloat res = p->object.callMethod<jfloat>(__FUNCTION__, "()F" );
    return res;
}

QString AsemanJavaLayer::packageName()
{
    QString res = p->object.callObjectMethod(__FUNCTION__, "()Ljava/lang/String;" ).toString();
    return res;
}

QString AsemanJavaLayer::deviceName()
{
    QString res = p->object.callObjectMethod(__FUNCTION__, "()Ljava/lang/String;" ).toString();
    return res;
}

QString AsemanJavaLayer::deviceId()
{
    QString res = p->object.callObjectMethod(__FUNCTION__, "()Ljava/lang/String;" ).toString();
    return res;
}

QRect AsemanJavaLayer::keyboardRect()
{
    jint jheight = p->object.callMethod<jfloat>("menuHeight", "()I" );
    int menuheight = jheight;

    const QList<QScreen*> & screens = QGuiApplication::screens();
    if( screens.isEmpty() )
        return QRect();

    QScreen *screen = screens.first();
    QRect rect = screen->availableGeometry();
    QRect geom = screen->geometry();

    rect.moveTop(rect.top() + menuheight);
    geom.setTop(geom.top() + menuheight);

    QRect final;

    if (rect != geom)
    {
        int ftop, fleft, fwidth, fheight;

        geom.getRect(&fleft, &ftop, &fwidth, &fheight);

        if (rect.top() != ftop)
            fheight = rect.top();
        else if (rect.left() != fleft)
            fwidth = rect.left();
        else if (rect.height() != fheight)
            ftop = rect.height();
        else if (rect.width() != fwidth)
            fleft = rect.width();

        final = QRect(fleft, ftop, fwidth - fleft, fheight - ftop);
    }

    return final;
}

QString AsemanJavaLayer::getLastImages(qint32 offset, qint32 count)
{
    jint joffset = offset;
    jint jcount = count;
    QString res = p->object.callObjectMethod(__FUNCTION__, "(II)Ljava/lang/String;", joffset, jcount).toString();
    return res;
}

QString AsemanJavaLayer::getContactList()
{
    QString res = p->object.callObjectMethod(__FUNCTION__, "()Ljava/lang/String;").toString();
    return res;
}

void AsemanJavaLayer::setKeepScreenOn(bool stt)
{
    jboolean jstt = stt;
    p->object.callMethod<jboolean>(__FUNCTION__, "(Z)V", jstt );
}

void AsemanJavaLayer::reloadBuffer()
{
    load_buffer();
}

bool AsemanJavaLayer::setTransparentNavigationBar(bool stt)
{
    jboolean jstt = stt;
    return p->object.callMethod<jboolean>(__FUNCTION__, "(Z)Z", jstt );
}

bool AsemanJavaLayer::setTransparentStatusBar(bool stt)
{
    jboolean jstt = stt;
    return p->object.callMethod<jboolean>(__FUNCTION__, "(Z)Z", jstt );\
}

void AsemanJavaLayer::load_buffer()
{
    while( !java_layer_inc_share_buffer.isEmpty() )
    {
        const QPair<QString,QString> & pair = java_layer_inc_share_buffer.takeFirst();
        Q_EMIT incomingShare( pair.first, pair.second );
    }
    while( !java_layer_inc_deeplinks_buffer.isEmpty() )
    {
        const auto & link = java_layer_inc_deeplinks_buffer.takeFirst();
        Q_EMIT deepLinkReceived(link);
    }
}

void AsemanJavaLayer::setImplemented(bool stt)
{
    jboolean jstt = stt;
    p->object.callMethod<jboolean>(__FUNCTION__, "(Z)Z", jstt );
}

AsemanJavaLayer::~AsemanJavaLayer()
{
    java_layers_objects.remove(this);
    delete p;
}

static void noteRecieved( JNIEnv *env, jobject obj ,jstring title, jstring msg )
{
    Q_UNUSED(obj)
    jboolean a;
    jboolean b;
    const char *t = env->GetStringUTFChars(title,&a);
    const char *m = env->GetStringUTFChars(msg,&b);

    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->incomingShare( QString(t), QString(m) );

    if( java_layers_objects.isEmpty() )
        java_layer_inc_share_buffer << QPair<QString,QString>( QString(t), QString(m) );
}

static void deepLinkReceived( JNIEnv *env, jobject obj ,jstring link)
{
    Q_UNUSED(obj)
    jboolean a;
    const char *l = env->GetStringUTFChars(link,&a);

    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->deepLinkReceived(QString(l));

    if( java_layers_objects.isEmpty() )
        java_layer_inc_deeplinks_buffer << QString(l);
}

static void imageRecieved( JNIEnv *env, jobject obj ,jstring jpath )
{
    Q_UNUSED(obj)
    jboolean a;
    const char *p = env->GetStringUTFChars(jpath,&a);

    QString path = QString("/sdcard/Aseman/%1.jpeg").arg(QDateTime::currentDateTime().toMSecsSinceEpoch());
    QFile().rename(QString(p),path);

    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->incomingImage(path);

    if( java_layers_objects.isEmpty() )
        java_layer_inc_image_buffer << path;
}

static void selectImageResult( JNIEnv *env, jobject obj ,jstring path )
{
    Q_UNUSED(obj)
    jboolean a;
    const char *p = env->GetStringUTFChars(path,&a);

    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->selectImageResult( QString(p) );
}

static void activityPaused( JNIEnv *env, jobject obj )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->activityPaused();
}

static void activityStopped( JNIEnv *env, jobject obj )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->activityStopped();
}

static void activityResumed( JNIEnv *env, jobject obj )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->activityResumed();
}

static void activityStarted( JNIEnv *env, jobject obj )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->activityStarted();
}

static void activityRestarted( JNIEnv *env, jobject obj )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->activityRestarted();
}

static void activityDestroyed( JNIEnv *env, jobject obj )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->activityDestroyed();
}

static void keyboardVisiblityChanged( JNIEnv *env, jobject obj, jint height )
{
    Q_UNUSED(env)
    Q_UNUSED(obj)
    for(AsemanJavaLayer *sjl: java_layers_objects)
        Q_EMIT sjl->keyboardVisiblityChanged(height);
}

bool aseman_jlayer_registerNativeMethods() {
    if(aseman_jlayer_native_methods_registered)
        return true;

    JNINativeMethod methods[] {{"_sendNote", "(Ljava/lang/String;Ljava/lang/String;)V", reinterpret_cast<void *>(noteRecieved)},
                               {"_sendImage", "(Ljava/lang/String;)V", reinterpret_cast<void *>(imageRecieved)},
                               {"_sendDeepLink", "(Ljava/lang/String;)V", reinterpret_cast<void *>(deepLinkReceived)},
                               {"_selectImageResult", "(Ljava/lang/String;)V", reinterpret_cast<void *>(selectImageResult)},
                               {"_activityPaused", "()V", reinterpret_cast<void *>(activityPaused)},
                               {"_activityStopped", "()V", reinterpret_cast<void *>(activityStopped)},
                               {"_activityResumed", "()V", reinterpret_cast<void *>(activityResumed)},
                               {"_activityStarted", "()V", reinterpret_cast<void *>(activityStarted)},
                               {"_activityRestarted", "()V", reinterpret_cast<void *>(activityRestarted)},
                               {"_activityDestroyed", "()V", reinterpret_cast<void *>(activityDestroyed)},
                               {"_keyboardVisiblityChanged", "(I)V", reinterpret_cast<void *>(keyboardVisiblityChanged)}};

    QAndroidJniObject javaClass("io/aseman/android/AsemanJavaLayer");
    QAndroidJniEnvironment env;
    jclass objectClass = env->GetObjectClass(javaClass.object<jobject>());

    env->RegisterNatives(objectClass, methods, sizeof(methods) / sizeof(methods[0]));

    env->DeleteLocalRef(objectClass);
    return true;
}

File: src/android/java/src/land/aseman/android/AsemanJavaLayer.java
/*
    Copyright (C) 2017 Aseman Team
    http://aseman.co

    AsemanQtTools is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AsemanQtTools is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package land.aseman.android;

import land.aseman.android.AsemanApplication;
import land.aseman.android.AsemanActivity;
import land.aseman.android.AsemanQtService;

import org.qtproject.qt5.android.bindings.QtService;

import android.util.Log;
import android.content.Intent;
import android.content.Context;
import android.app.Activity;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.provider.MediaStore;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.text.TextUtils;
import android.view.WindowManager;
import android.os.Build;
import android.content.res.Configuration;
import android.view.View;
import android.view.Window;
import android.graphics.Rect;
import android.provider.Settings.Secure;
import java.lang.Runnable;
import android.os.Handler;

import java.io.File;
import java.io.InputStream;
import java.io.FileOutputStream;

import android.app.ActivityManager.RunningAppProcessInfo;
import android.app.ActivityManager ;
import android.os.Process;
import java.util.List;
import java.util.Iterator;

//import com.google.android.gms.auth.api.signin.GoogleSignInOptions;

public class AsemanJavaLayer
{
    private static native void _sendNote( String title, String msg );
    private static native void _sendImage( String path );
    private static native void _activityPaused();
    private static native void _activityStopped();
    private static native void _activityResumed();
    private static native void _activityStarted();
    private static native void _activityRestarted();
    private static native void _activityDestroyed();
    private static native void _selectImageResult( String path );
    private static native void _keyboardVisiblityChanged(int height);
    private static boolean implemented = false;

    public AsemanJavaLayer() {
    }

    public static void activityPaused(){
        if(implemented)
            _activityPaused();
    }

    public static void activityStopped(){
        if(implemented)
            _activityStopped();
    }

    public static void activityResumed(){
        if(implemented)
            _activityResumed();
    }

    public static void activityStarted(){
        if(implemented)
            _activityStarted();
    }

    public static void activityRestarted(){
        if(implemented)
            _activityRestarted();
    }

    public static void activityDestroyed(){
        if(implemented)
            _activityDestroyed();
    }

    public static void keyboardVisiblityChanged(int height){
        if(implemented)
            _keyboardVisiblityChanged(height);
    }

    public static void sendNote( String title, String msg ) {
        if( title == null )
            title = "";
        if( msg == null )
            msg = "";

        if(implemented)
            _sendNote(title,msg);
    }

    public static void setKeepScreenOn(boolean status) {
        final boolean fstatus = status;
        final AsemanActivity activtiy = AsemanActivity.getActivityInstance();
        Handler mainHandler = new Handler(activtiy.getMainLooper());
        Runnable myRunnable = new Runnable() {
            @Override
            public void run() {
                activtiy.setKeepScreenOn(fstatus);
            }
        };
        mainHandler.post(myRunnable);
    }

    public static int menuHeight()
    {
        Rect r = new Rect();
        Window window = AsemanActivity.getActivityInstance().getWindow();
        View rootview = window.getDecorView();
        rootview.getWindowVisibleDisplayFrame(r);

        return r.top;
    }

    public static void sendImage( Uri data ) {
        String imagePath = "/sdcard/Aseman/tmp_input_image.jpg";

        File file = new File(imagePath);
        file.delete();

        try {
            InputStream stream = AsemanApplication.instance().getContentResolver().openInputStream(data);
            Bitmap bitmap = BitmapFactory.decodeStream(stream);
            stream.close();

            FileOutputStream out = new FileOutputStream(imagePath);
            bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out);
            out.close();
        } catch(Exception e) {
            return;
        }

        if(implemented)
            _sendImage(imagePath);
    }

    public static void selectImageResult( String path ) {
        if( path == null )
            path = "";

        if(implemented)
            _selectImageResult(path);
    }

    public String packageName()
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();
        return oContext.getPackageName();
    }

    public String deviceName() {
        String manufacturer = Build.MANUFACTURER;
        String model = Build.MODEL;
        if (model.startsWith(manufacturer)) {
            return capitalize(model);
        } else {
            return capitalize(manufacturer) + " " + model;
        }
    }

    boolean setImplemented(boolean stt) {
        implemented = stt;
        return true;
    }

    public String deviceId() {
        return Secure.getString(getContext().getContentResolver(), Secure.ANDROID_ID);
    }


    private String capitalize(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        char first = s.charAt(0);
        if (Character.isUpperCase(first)) {
            return s;
        } else {
            return Character.toUpperCase(first) + s.substring(1);
        }
    }

    public static Context getContext() {
        if(AsemanActivity.getActivityInstance() != null)
            return AsemanActivity.getActivityInstance();
        else
        if(AsemanQtService.getServiceInstance() != null)
            return AsemanQtService.getServiceInstance();
        else
            return AsemanApplication.getAppContext();
    }

    boolean killService(String serviceName)
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();
        ActivityManager am = (ActivityManager) oContext.getSystemService(oContext.ACTIVITY_SERVICE);
        List<RunningAppProcessInfo> runningAppProcesses = am.getRunningAppProcesses();

        Iterator<RunningAppProcessInfo> iter = runningAppProcesses.iterator();

        while(iter.hasNext()){
            RunningAppProcessInfo next = iter.next();

            String pricessName = oContext.getPackageName() + serviceName;

            if(next.processName.equals(pricessName)){
                Process.killProcess(next.pid);
                break;
            }
        }

        return true;
    }

    boolean startQtService()
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();
        Intent i = new Intent(oContext, AsemanQtService.class);
        i.putExtra("name", "SurvivingwithAndroid");
        try {
            oContext.startService(i);
        } catch(Exception e) {
            return false;
        }

        return true;
    }

    boolean stopQtService()
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();
        Intent i = new Intent(oContext, AsemanQtService.class);
        try {
        oContext.stopService(i);
        } catch(Exception e) {
            return false;
        }
        return true;
    }

    boolean sharePaper( String title, String msg )
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();

        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TEXT, msg);
        intent.putExtra(Intent.EXTRA_SUBJECT, title);

        try {
            oContext.startActivity(intent);
        } catch(Exception e) {
            return false;
        }
        return true;
    }

    boolean shareFile(String path, String type)
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();

        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setType(type);
        intent.putExtra(Intent.EXTRA_STREAM, Uri.parse("file://"+path));

        try {

            Intent new_intent = Intent.createChooser(intent , "Share Via");
            new_intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            oContext.startActivity(new_intent);

        } catch(Exception e) {
            Log.e("Share" , "Exception:" , e);
            return false;
        }
        return true;
    }


    boolean openFile( String path, String type )
    {
        Context oContext;
        oContext = AsemanApplication.getAppContext();

        Intent intent = new Intent(android.content.Intent.ACTION_VIEW);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setType(type);
        intent.setDataAndType(Uri.parse(path),type);

        try {
            oContext.startActivity(intent);
        } catch(Exception e) {
            return false;
        }
        return true;
    }

    DisplayMetrics getMetrics()
    {
        DisplayMetrics dm = new DisplayMetrics();
        try {
            AsemanActivity.getActivityInstance().getWindowManager().getDefaultDisplay().getMetrics(dm);
        } catch(Exception e) {
            return dm;
        }
        return dm;
    }

    double lcdDpiX()
    {
        return getMetrics().densityDpi;
    }

    int densityDpi()
    {
        return getMetrics().densityDpi;
    }

    float density()
    {
        return getMetrics().density;
    }

    int screenSizeWidth()
    {
        return getMetrics().widthPixels;
    }

    int screenSizeHeight()
    {
        return getMetrics().heightPixels;
    }

    int statusBarHeight() {
        int result = 0;
        Context context = AsemanApplication.getAppContext();
        int resourceId = context.getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = context.getResources().getDimensionPixelSize(resourceId);
        }
        return result;
    }

    int navigationBarHeight() {
        int result = 0;
        Context context = AsemanApplication.getAppContext();
        int resourceId = context.getResources().getIdentifier("navigation_bar_height", "dimen", "android");
        if (resourceId > 0) {
            return context.getResources().getDimensionPixelSize(resourceId);
        }
        return 0;
    }

    boolean isTablet()
    {
        Context context = AsemanApplication.getAppContext();
        return (context.getResources().getConfiguration().screenLayout
                & Configuration.SCREENLAYOUT_SIZE_MASK)
                >= Configuration.SCREENLAYOUT_SIZE_LARGE;
    }

    int getSizeName()
    {
        Context context = AsemanApplication.getAppContext();
        int screenLayout = context.getResources().getConfiguration().screenLayout;
        screenLayout &= Configuration.SCREENLAYOUT_SIZE_MASK;

        switch (screenLayout) {
        case Configuration.SCREENLAYOUT_SIZE_SMALL:
            return 0;
        case Configuration.SCREENLAYOUT_SIZE_NORMAL:
            return 1;
        case Configuration.SCREENLAYOUT_SIZE_LARGE:
            return 2;
        case 4: // Configuration.SCREENLAYOUT_SIZE_XLARGE is API >= 9
            return 3;
        default:
            return -1;
        }
    }

    boolean getOpenPictures()
    {
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("image/*");

        try {
            AsemanActivity.getActivityInstance().startActivityForResult(intent, AsemanActivity.SELECT_IMAGE);
        } catch(Exception e) {
            return false;
        }

        return true;
    }

    boolean startCamera( String output )
    {
        Intent intent = new Intent("android.media.action.IMAGE_CAPTURE");
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(new File(output)));

        try {
            AsemanActivity.getActivityInstance().startActivityForResult(intent, 0);
        } catch(Exception e) {
            return false;
        }
        return true;
    }

    boolean callNumber( String number )
    {
        Intent intent = new Intent(Intent.ACTION_CALL);
        intent.setData(Uri.parse("tel:" + number));

        try {
            AsemanActivity.getActivityInstance().startActivity(intent);
        } catch(Exception e) {
            return false;
        }
        return true;
    }

    boolean transparentStatusBar()
    {
        return AsemanActivity.getActivityInstance().transparentStatusBar();
    }

    boolean transparentNavigationBar()
    {
        return AsemanActivity.getActivityInstance().transparentNavigationBar();
    }

    boolean release()
    {
        return true;
    }

    boolean setTransparentNavigationBar(boolean stt)
    {
        final boolean _stt = stt;
        Handler mainHandler = new Handler(AsemanActivity.getActivityInstance().getMainLooper());
        Runnable myRunnable = new Runnable() {
            @Override
            public void run() {
                AsemanActivity.getActivityInstance().setTransparentNavigationBar(_stt);
            }
        };
        mainHandler.post(myRunnable);
        return true;
    }

    boolean setTransparentStatusBar(boolean stt)
    {
        final boolean _stt = stt;
        Handler mainHandler = new Handler(AsemanActivity.getActivityInstance().getMainLooper());
        Runnable myRunnable = new Runnable() {
            @Override
            public void run() {
                AsemanActivity.getActivityInstance().setTransparentStatusBar(_stt);
            }
        };
        mainHandler.post(myRunnable);
        return true;
    }

    boolean startForeground(int id, String title, String body, String iconPath, String icon)
    {
        if(AsemanQtService.getServiceInstance() != null)
            return AsemanQtService.getServiceInstance().startForeground(id, title, body, iconPath, icon);
        return false;
    }

    boolean stopForeground(boolean removeNotification)
    {
        if(AsemanQtService.getServiceInstance() != null) {
            AsemanQtService.getServiceInstance().stopForeground(removeNotification);
            return true;
        }
        return false;
    }

    boolean startNotification(int id, String title, String body, String iconPath, String icon)
    {
        if(AsemanActivity.getActivityInstance() != null)
            return AsemanActivity.getActivityInstance().startNotification(id, title, body, iconPath, icon);
        else
        if(AsemanQtService.getServiceInstance() != null)
            return AsemanQtService.getServiceInstance().startNotification(id, title, body, iconPath, icon);
        return false;
    }

    boolean stopNotification(int id)
    {
        if(AsemanActivity.getActivityInstance() != null)
            return AsemanActivity.getActivityInstance().stopNotification(id);
        else
        if(AsemanQtService.getServiceInstance() != null)
            return AsemanQtService.getServiceInstance().stopNotification(id);
        return false;
    }

//    public static String getUniquePsuedoID() {
//        String m_szDevIDShort = "35" + (Build.BOARD.length() % 10) + (Build.BRAND.length() % 10) + (Build.CPU_ABI.length() % 10) + (Build.DEVICE.length() % 10) + (Build.MANUFACTURER.length() % 10) + (Build.MODEL.length() % 10) + (Build.PRODUCT.length() % 10);
//        String serial = null;
//        try {
//            serial = android.os.Build.class.getField("SERIAL").get(null).toString();
//            return new UUID(m_szDevIDShort.hashCode(), serial.hashCode()).toString();
//        } catch (Exception exception) {
//            serial = "3077139d-9dde-4ee1-afb5-ecc43f5d1524";
//        }
//        return new UUID(m_szDevIDShort.hashCode(), serial.hashCode()).toString();
//    }

//    boolean googleSignIn(String clientId)
//    {
//        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
//            .requestIdToken(getString(clientId))
//            .build();
//    }
}
